docs/src/design/proposals/APIRedesign.md

```diff
@@ -0,0 +1,1006 @@
+recognize which constructor arguments are from dependencies, and
+which will be provided at run-time. Instead, it is the responsibility
+of whoever is introducing a component to manage these things
+separately.
+
+### Benefits
+
+* More clarity.
+
+### Detriments
+
+* Possibly results in redundant effort to manage this difference
+  (other APIs may need to be adjusted accordingly.)
+
+## Rename Views to Applications
+
+Rename (internally to the application, not necessarily in UI or
+in user guide) what are currently called `views` to `applications`.
+
+### Benefits
+
+* Easier to understand. What is currently called a "view" is,
+  in the MVC sense, a view-controller pair, usually with its own
+  internal model for view state. Calling these "applications"
+  would avoid this ambiguity/inconsistency.
+* Also provides an appropriate mindset for building these;
+  particularly, sets the expectation that you'll want to decompose
+  this "application" into smaller pieces. This nudges developers
+  in appropriate directions (in contrast to `views`, which
+  typically get implemented as templates with over-complicated
+  "controllers".)
+
+### Detriments
+
+* Developer terminology falls out of sync with what is used in
+  the user guide.
+
+## Provide Classes for Extensions
+
+As a general pattern, when introducing extension categories, provide
+classes with a standard implementation of these interfaces that
+plugin developers can `new` and register.
+
+For example, instead of declaring a type as:
+
+```json
+{
+  "types": [{
+    "key": "sometype",
+    "glyph": "X",
+    "etc": "..."
+  }]
+}
+```
+
+You would register one as:
+
+```js
+// Assume we have gotten a reference to a type registry somehow
+typeRegistry.register(new Type({
+    "key": "sometype",
+    "glyph": "X",
+    "etc": "..."
+}));
+```
+
+### Benefits
+
+* Easier to understand (less "magic").
+* Lends itself naturally to substitution of different implementations
+  of the same interface.
+* Allows for run-time decisions about exactly what gets registered.
+
+### Detriments
+
+* Adds some modest boilerplate.
+* Provides more opportunity to "do it wrong."
+
+## Normalize naming conventions
+
+Adopt and obey the following naming conventions for AMD modules
+(and for injectable dependencies, which may end up being modules):
+
+* Use `UpperCamelCase` for classes.
+* Use `lowerCase` names for instances.
+    * Use `someNoun` for object instances which implement some
+      interface. The noun should match the implemented interface,
+      when applicable.
+    * `useSomeVerb` for functions.
+* Use `ALL_CAPS_WITH_UNDERSCORES` for other values, including
+  "struct-like" objects (that is, where the object conceptually
+  contains properties rather than methods.)
+
+### Benefits
+
+* Once familiar with the conventions, easier to understand what
+  individual modules are.
+
+### Detriments
+
+* A little bit inflexible.
+
+## Expose no third-party APIs
```

Comment:
Thanks for bringing this discussion over.

> Whether you're using require, browserify or plain script tags, the concept of a dependency loader is something you should be familiar with in every language. People not interested in gaining a cursory understanding of these tools are really saying "I don't want to learn anything about web technologies". Do we really want to cater to these people?

The people we need to cater to are GDS developers. In that context, the learning curve matters in two ways:

* During a trade study, they will need to estimate the effort to integrate and maintain our system in whatever configuration their mission needs.
* During development, they will need to achieve meaningful mission-specific results (integration with data) in a reasonable time frame. 

In this context, it's extremely common to encounter people who have a lot of domain knowledge, as opposed to knowledge about any specific language/platform. They are going to have to learn A, B, and C; there's not much we can do to prevent that. But we can remove, reduce or postpone  the need to learn some of these other technologies and make things easier that way. 

Right now, you need to understand how Require works, how Angular works, and how Open MCT Web works, all before you can begin writing plugins. It's completely feasible for us to provide an API that allows developers to just include us with a script tag and say `mct.extend(inThisWay)`, so why make it more complicated than that?

> When people say "I have to learn X, Y, and Z to use this" and then claim that they'd prefer to say "I only have to learn Z," well, if Z included all of X, Y, and Z, then they'd still have to learn it all.

To reiterate, I'm not stating we should wrap all functionality of all libraries we use. RequireJS is useful to us so we should use it. Do we need to expose it or wrap it? No; Open MCT Web is not a script loader, so we shouldn't need to provide any script-loading capability. (This is a case where I think the current API and bundle-loading mechanism overreach.) Lodash/Underscore are great "utility belts"; Open MCT Web isn't there to provide you with a utility belt, no need for us to expose that. 

In other words, Z doesn't include all of X and Y. If it did, I'd agree that this would be silly. But we're only ever using a subset of these libraries, and the subset that we actually need to _expose_ is smaller still. I'm saying let's just let Z be Z. The functionality that we expose (that we supply an API for) can and should be completely orthogonal to that of our dependencies. We're not adding any value by exposing an API that people could get and use themselves, we're only taking on a liability.

The one dependency where I see this getting at all complicated/non-trivial is with Angular. Obviously, we don't want to "simply wrap" all of the utility services it exposes, or its template syntax. There's a case to be made there for Open MCT Web just being another Angular module (or set of modules); in that context, we're not really exposing Angular APIs, we're just assuming those APIs are present. However, let's also consider that there are good reasons for [wanting to reduce our dependency on Angular](https://github.com/nasa/openmctweb/blob/api-redesign/docs/src/design/proposals/APIRedesign.md#angularjs) that are totally separate from this. 

As an example for where I think wrapping would be useful: We've gotten feedback from developers that they'd like to write [views without Angular](https://github.com/nasa/openmctweb/blob/api-redesign/docs/src/design/proposals/APIRedesign.md#arbitrary-html-views); if we accepted an interface for views that just communicated plain-old-HTML elements, we could internally use an implementation of that which trivially wraps compiled Angular templates, and expose that implementation for others to use through a non-platform plug-in. We could also wrap services like `$http` or `$timeout` - or we could just say we're not responsible for providing that kind of functionality.

> As much as we call openmctweb a platform, I would better define it as a set of tools, common interfaces, and design patterns that is intended to be easily extensible and promote good development practices.

I have a preference for [extensible application](https://github.com/nasa/openmctweb/blob/api-redesign/docs/src/design/proposals/APIRedesign.md#nomenclature-change), but yes, same idea.